celements-spring-security/src/main/java/com/celements/spring/security/OAuth2CookieService.java

```diff
@@ -0,0 +1,163 @@
+            // here's where you hand Spring your pre‐built client:
+            .attribute(OAuth2AuthorizedClient.class.getName(), existingClient)
+            .build())
+        .map(authorizedClientManager::authorize)
+        .filter(client -> !client.getAccessToken().getTokenValue()
+            .equals(oldClientOpt.get().getAccessToken().getTokenValue())
+            || ((client.getRefreshToken() != null) && !client.getRefreshToken().getTokenValue()
+                .equals(oldClientOpt.get().getRefreshToken().getTokenValue())));
+  }
+
+  public Optional<OAuth2AuthorizedClient> reconstructAuthClientFromCookie(HttpServletRequest req) {
+    try {
+      Optional<Jwt> accessJwtOpt = getJwtFromCookie(req, COOKIE_ACCESS_TOKEN);
+      Optional<OAuth2RefreshToken> refreshTokenOpt = getRefreshTokenFromCookie(req);
+      if (accessJwtOpt.isPresent() && refreshTokenOpt.isPresent()) {
+        OAuth2AccessToken accessToken = reconstructAccessTokenFromJwt(accessJwtOpt.get());
+        OAuth2RefreshToken refreshToken = refreshTokenOpt.get();
+        return Optional.of(new OAuth2AuthorizedClient(
+            registrationRepo.findByRegistrationId(identityService.getRegistrationId()),
+            accessJwtOpt.get().getSubject(),
+            accessToken,
+            refreshToken));
+      }
+    } catch (OAuth2AuthenticationException exp) {
+      LOGGER.info("No valid access token found", exp);
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * @param request
+   *          the request
+   * @return the Refresh Token or optional.empty if none found
+   * @throws OAuth2AuthenticationException
+   *           if the found token is invalid
+   */
+  public Optional<OAuth2RefreshToken> getRefreshTokenFromCookie(HttpServletRequest req) {
+    return getJwtFromCookie(req, COOKIE_REFRESH_TOKEN)
+        .map(refreshToken -> new OAuth2RefreshToken(
+            refreshToken.getTokenValue(),
+            refreshToken.getIssuedAt()));
+  }
+
+  /**
+   * Resolve any
+   * <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer
+   * Token</a> value from the access_token cookie in the request.
+   *
+   * @param request
+   *          the request
+   * @return the Bearer Access Token or optional.empty if none found
+   * @throws OAuth2AuthenticationException
+   *           if the found token is invalid
+   */
+  public Optional<OAuth2AccessToken> getAccessTokenFromCookie(HttpServletRequest req) {
+    return getJwtFromCookie(req, COOKIE_ACCESS_TOKEN)
+        .map(this::reconstructAccessTokenFromJwt);
+  }
+
+  private OAuth2AccessToken reconstructAccessTokenFromJwt(Jwt accessToken) {
+    return new OAuth2AccessToken(
+        OAuth2AccessToken.TokenType.BEARER,
+        accessToken.getTokenValue(),
+        accessToken.getIssuedAt(),
+        accessToken.getExpiresAt(),
+        getScopesFromJwt(accessToken));
+  }
+
+  public void setTokenCookie(HttpServletResponse response, String name, String value,
```

Comment:
```suggestion
  private void setTokenCookie(HttpServletResponse response, String name, String value, Instant expiry) {
```
make private and add import